ChopChop/ViewModels/RecipeFormViewModel.swift

```diff
@@ -0,0 +1,214 @@
+//
+//  RecipeFormViewModel.swift
+//  ChopChop
+//
+//  Created by Cao Wenjie on 20/3/21.
+//
+import SwiftUI
+import Combine
+import GRDB
+
+class RecipeFormViewModel: ObservableObject {
+
+    private var existingRecipe: Recipe?
+    @Published var hasError: Bool = false
+    private var recipeId: Int64?
+    private(set) var errorMessage = ""
+    private(set) var isEdit = false
+    private let storageManager = StorageManager()
+    private var recipeCategoryCancellable = Set<AnyCancellable>()
+    @Published var recipeName = ""
+    @Published var serving = "" {
+        didSet {
+            ensureValidServing()
+        }
+    }
+    @Published var allRecipeCategories = [RecipeCategory]()
+    @Published var recipeCategory = ""
+    @Published var difficulty: Difficulty = .veryEasy
+    private var storage = StorageManager()
+    @Published var steps = [String]()
+    @Published var ingredients = [RecipeIngredientRowViewModel]()
+    @Published var ingredientParsingString = ""
+    @Published var instructionParsingString = ""
+
+    init(recipe: Recipe) {
+        existingRecipe = recipe
+        recipeId = recipe.id
+        recipeName = recipe.name
+        serving = recipe.servings.description
+        difficulty = recipe.difficulty ?? .veryEasy
+
+        steps = recipe.steps.map({ $0.content })
+        ingredients = recipe.ingredients.map({
+            RecipeIngredientRowViewModel(
+                amount: $0.quantity.value.description,
+                unit: $0.quantity.type,
+                ingredientName: $0.name
+            )
+        })
+        fetchCategories()
+        isEdit = true
+    }
+
+    init() {
+        fetchCategories()
+    }
+
+    private func fetchCategories() {
+        storageManager
+            .recipeCategoriesPublisher()
+            .receive(on: DispatchQueue.main)
+            .sink(receiveCompletion: { [weak self] value in
+                guard let self = self else {
+                    return
+                }
+                switch value {
+                case .failure:
+                    self.allRecipeCategories = []
+                case .finished:
+                    break
+                }
+            },
+            receiveValue: { [weak self] categories in
+                guard let self = self else {
+                    return
+                }
+                self.allRecipeCategories = categories
+            })
+            .store(in: &recipeCategoryCancellable)
+    }
+
+    func ensureValidServing() {
+        let filtered = serving.filter { "0123456789.".contains($0) }
+        if filtered != serving {
+            serving = filtered
+        }
+    }
+
+    func checkFormValid() throws {
+        if recipeName.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty {
```

Comment:
by right the models are supposed to help you with these?

Reply:
The model can only handle the name part so I thought it will be more consistent to do all the checking at the form level instead of some at form some at model. The FormError thrown will also be more appropriate than the error thrown in model.

Reply:
Can update the RecipeModel to be able to handle these checks?